### PR TITLE
tidb-lightning: max-error and duplicate-resolution

### PR DESCRIPTION
**Pull request by hfxsd**
### What is changed, added or deleted? (Required)

Document the error resolution features, introduced in pingcap/tidb#27181 and pingcap/tidb#28464.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):https://github.com/pingcap/docs-cn/pull/7430

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 1]
```diff
@@ -0,0 +1,308 @@
+---
+title: TiDB Lightning 错误处理功能
+summary: 介绍了如何解决导入数据过程中的类型转换和冲突错误。
+---
+
+# TiDB Lightning 错误处理功能
+
+> **警告:**
+>
+> TiDB Lightning 错误处理功能是实验特性。**不建议**在生产环境中仅依赖该功能处理相关错误。
```
Comment (gozssky) [inline tidb-lightning/tidb-lightning-error-resolution.md:10]: @glorv @sunzhaoyang 错误处理在 5.4 还是实验特性吗

Comment (glorv) [reply tidb-lightning/tidb-lightning-error-resolution.md:10]: 是的

**Review by gozssky — APPROVED** [review]

**Review by glorv — APPROVED** [review]
LGTM

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+从 TiDB 5.3.0 开始，你可以配置 TiDB Lightning 以跳过诸如无效类型转换、唯一键冲突等错误，让导入任务持续进行，就如同出现错误的行数据不存在一样。你可以依据生成的报告，手动修复这些错误。该功能适用于以下场景：
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:11]:
```suggestion
从 TiDB 5.4.0 开始，你可以配置 TiDB Lightning 以跳过诸如无效类型转换、唯一键冲突等错误，让导入任务持续进行，就如同出现错误的行数据不存在一样。你可以依据生成的报告，手动修复这些错误。该功能适用于以下场景：
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+     table_name: `example`.`t`
+           path: example.t.1.sql
+         offset: 253
+          error: failed to cast value as tinyint(4) for column `a` (#1): [types:1690]constant 600 overflows tinyint
+       row_data: (600,'six hundred')
+    ```
+
+7. 检查 `conflict_error_v1` 表是否捕获了具有唯一键/主键冲突的四行：
+
+    ```console
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:262]:
```suggestion
    ```sql
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+    | a | b   |
+    +---+-----+
+    | 1 | one |
+    | 2 | two |
+    +---+-----+
+    ```
+
+6. 检查 `type_error_v1` 表是否捕获了涉及类型转换的三行：
+
+    ```console
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:232]:
```suggestion
    ```sql
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+    mkdir example && cd example
+    echo 'CREATE SCHEMA example;' > example-schema-create.sql
+    echo 'CREATE TABLE t(a TINYINT PRIMARY KEY, b VARCHAR(12) NOT NULL UNIQUE);' > example.t-schema.sql
+    ```
+
+2. 准备数据：
+
+    {{< copyable "shell-regular" >}}
+
+    ```sh
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:172]:
```suggestion
    ```shell
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+下列错误是致命错误，不能通过配置 `max-error` 跳过：
+
+* 原始 CSV、SQL 或者 Parquet 文件中的语法错误，例如未闭合的引号
+* I/O、网络、或系统权限错误
+
+在 Local 后端模式下，唯一键/主键的冲突是单独处理的。相关内容将在接下来的章节进行介绍。
+
+## 解决重复问题
+
+Local 后端模式下，TiDB Lightning 导入数据时先将数据转换成 KV 对数组（KV pairs），然后批量添加到 TiKV 中。与 TiDB 后端模式不同，直到任务结束才会检测重复行。因此，Local 模式下的重复错误不是通过 `max-error` 进行控制，而是通过 `duplicate-resolution` 配置项进行控制的。
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:49]:
```suggestion
Local-backend 模式下，TiDB Lightning 导入数据时先将数据转换成 KV 对数组（KV pairs），然后批量添加到 TiKV 中。与 TiDB-backend 模式不同，TiDB Lightning 在 Local-backend 模式下直到任务结束才会检测重复行。因此，Local 模式下的重复错误不是通过 `max-error` 进行控制，而是通过 `duplicate-resolution` 配置项进行控制的。
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+```toml
+[tikv-importer]
+duplicate-resolution = 'none'
+```
+
+`duplicate-resolution` 有以下三个选项：
+
+* **'none'**：不对重复数据进行检测。如果唯一键/主键冲突确实存在，那么导入的表格里会出现不一致的数据和索引，checksum 检查的时候会失败。
+* **'record'**：检测重复数据，但不会对重复数据进行修复。如果唯一键/主键冲突确实存在，那么导入的表格里会出现不一致的数据和索引，checksum 检查的时候会失败。
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:61]:
```suggestion
* **'none'**：不对重复数据进行检测。如果唯一键或主键冲突确实存在，那么导入的表格里会出现不一致的数据和索引，checksum 检查的时候会失败。
* **'record'**：检测重复数据，但不会对重复数据进行修复。如果唯一键或主键冲突确实存在，那么导入的表格里会出现不一致的数据和索引，checksum 检查的时候会失败。
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+> **警告:**
+>
+> TiDB Lightning 错误处理功能是实验特性。**不建议**在生产环境中仅依赖该功能处理相关错误。
+从 TiDB 5.3.0 开始，你可以配置 TiDB Lightning 以跳过诸如无效类型转换、唯一键冲突等错误，让导入任务持续进行，就如同出现错误的行数据不存在一样。你可以依据生成的报告，手动修复这些错误。该功能适用于以下场景：
+
+- 要导入的数据有少许错误
+- 手动定位错误比较困难
+- 如果遇到错误就重启 TiDB Lightning，代价太大
+
+## 类型错误 (Type error)
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:17]: "类型错误"、"解决重复问题"和"错误报告"三部分的前后衔接不是很清晰，不能很直观地看出这三部分在错误处理这个特性中有什么作用。能否在各部分开头添加一下描述呢？

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+         offset: 46
+          error: failed to cast value as varchar(12) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin for column `b` (#2): [table:1048]Column 'b' cannot be null
+       row_data: (0,NULL)
+    *************************** 2. row ***************************
+        task_id: 1635888701843303564
+    create_time: 2021-11-02 21:31:42.627496
+     table_name: `example`.`t`
+           path: example.t.1.sql
+         offset: 183
+          error: failed to cast value as varchar(12) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin for column `b` (#2): [types:1406]Data Too Long, field len 12, data len 13
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:248]: 这个特性是否有使用限制呢？

Comment (glorv) [reply tidb-lightning/tidb-lightning-error-resolution.md:248]: 应该没什么限制

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+* 其他 SQL 错误
+
+下列错误是致命错误，不能通过配置 `max-error` 跳过：
+
+* 原始 CSV、SQL 或者 Parquet 文件中的语法错误，例如未闭合的引号
+* I/O、网络、或系统权限错误
+
+在 Local 后端模式下，唯一键/主键的冲突是单独处理的。相关内容将在接下来的章节进行介绍。
+
+## 解决重复问题
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:47]: @kennytm 这里是GA，max-error 又是实验特性。是否把两个区分下？

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+        (40, 'fourty'),         -- 与上面的 `40` 冲突
+        (42, 'fifty-four');     -- 与上面的 `'fifty-four'` 冲突
+    EOF
+    ```
+
+3. 配置 TiDB Lightning，启用严格 SQL 模式，使用 Local 后端模式进行导入，通过删除解决重复项，并最多跳过 10 个错误：
+
+    {{< copyable "shell-regular" >}}
+
+    ```sh
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:191]:
```suggestion
    ```shell
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+        password = ''
+        sql-mode = 'STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE'
+    EOF
+    ```
+
+4. 运行 TiDB Lightning。因为已跳过所有错误，该命令执行完会成功退出：
+
+    {{< copyable "shell-regular" >}}
+
+    ```sh
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:214]:
```suggestion
    ```shell
```

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 8 above)
Comment (hfxsd) [reply tidb-lightning/tidb-lightning-error-resolution.md:14]: @gozssky PTAL, thanks!

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+
+- 要导入的数据有少许错误
+- 手动定位错误比较困难
+- 如果遇到错误就重启 TiDB Lightning，代价太大
+
+## 类型错误 (Type error)
+
+> **警告:**
+>
+> TiDB Lightning 类型错误处理功能（`lightning.max-error`）是实验特性。**不建议**在生产环境中仅依赖该功能处理相关错误。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:18]: 在这里加了个警告。PLAL @gozssky

**Review by gozssky — APPROVED** [review]

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+| raw_key      |        |      | ✓        | 冲突的 KV 对的键                                                           |
+| raw_value    |        |      | ✓        | 冲突的 KV 对的值                                                            |
+| raw_handle   |        |      | ✓        | 冲突行的行句柄                                                         |
+| raw_row      |        |      | ✓        | 冲突行的编码值                                                       |
+
+> **注意：**
+>
+> 错误报告记录的是文件偏移量，不是行号或列号，因为行号或列号的获取效率很低。你可以使用下列命令在字节位实现快速跳转（以 183 为例）：
+>
+> * shell, 输出前面几行——
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:141]:
```suggestion
> * shell：输出前面几行
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+>
+>     ```sh
+>     head -c 183 file.csv | tail
+>     ```
+>
+> * shell, 输出后面几行——
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:147]:
```suggestion
> * shell，输出后面几行
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+>
+>     ```sh
+>     tail -c +183 file.csv | head
+>     ```
+>
+> * vim — `:goto 183` 或 `183go`
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:153]:
```suggestion
> * vim：`:goto 183` 或 `183go`
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+| raw_handle   |        |      | ✓        | 冲突行的行句柄                                                         |
+| raw_row      |        |      | ✓        | 冲突行的编码值                                                       |
+
+> **注意：**
+>
+> 错误报告记录的是文件偏移量，不是行号或列号，因为行号或列号的获取效率很低。你可以使用下列命令在字节位实现快速跳转（以 183 为例）：
+>
+> * shell, 输出前面几行——
+>
+>     ```sh
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:143]:
```suggestion
>     ```shell
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+>     head -c 183 file.csv | tail
+>     ```
+>
+> * shell, 输出后面几行——
+>
+>     ```sh
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:149]:
```suggestion
>     ```shell
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+## 错误报告
+
+{{< copyable "" >}}
+
+```toml
+[lightning]
+task-info-schema-name = 'lightning_task_info'
+```
+
+所有错误都会写入下游 TiDB 集群 `lightning_task_info` 数据库中的表中。你可以使用 `lightning.task-info-schema-name` 配置更改数据库名称。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:76]:
````suggestion
所有错误都会写入下游 TiDB 集群 `lightning_task_info` 数据库中的表中。在导入完成后，你可以根据数据库中记录的内容，手动进行处理。

你可以使用 `lightning.task-info-schema-name` 配置更改数据库名称。

{{< copyable "" >}}

```toml
[lightning]
task-info-schema-name = 'lightning_task_info'
```

````

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+下列错误是致命错误，不能通过配置 `max-error` 跳过：
+
+* 原始 CSV、SQL 或者 Parquet 文件中的语法错误，例如未闭合的引号
+* I/O、网络、或系统权限错误
+
+在 Local 后端模式下，唯一键/主键的冲突是单独处理的。相关内容将在接下来的章节进行介绍。
+
+## 解决重复问题
+
+Local-backend 模式下，TiDB Lightning 导入数据时先将数据转换成 KV 对数组（KV pairs），然后批量添加到 TiKV 中。与 TiDB-backend 模式不同，TiDB Lightning 在 Local-backend 模式下直到任务结束才会检测重复行。因此，Local 模式下的重复错误不是通过 `max-error` 进行控制，而是通过 `duplicate-resolution` 配置项进行控制的。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:50]:
```suggestion
Local-backend 模式下，TiDB Lightning 导入数据时先将数据转换成 KV 对数组（KV pairs），然后批量添加到 TiKV 中。与 TiDB-backend 模式不同，TiDB Lightning 在 Local-backend 模式下直到任务结束才会检测重复行。因此，Local-backend 模式下的重复错误不是通过 `max-error` 进行控制，而是通过 `duplicate-resolution` 配置项进行控制的。你可以通过配置该参数的行为，来决定如何处理有冲突的数据。
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+* 其他 SQL 错误
+
+下列错误是致命错误，不能通过配置 `max-error` 跳过：
+
+* 原始 CSV、SQL 或者 Parquet 文件中的语法错误，例如未闭合的引号
+* I/O、网络、或系统权限错误
+
+在 Local 后端模式下，唯一键/主键的冲突是单独处理的。相关内容将在接下来的章节进行介绍。
+
+## 解决重复问题
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:48]:
```suggestion
## Local-backend 模式下解决重复问题
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+- 如果遇到错误就重启 TiDB Lightning，代价太大
+
+## 类型错误 (Type error)
+
+> **警告:**
+>
+> TiDB Lightning 类型错误处理功能（`lightning.max-error`）是实验特性。**不建议**在生产环境中仅依赖该功能处理相关错误。
+
+你可以通过修改配置项 `lightning.max-error` 来增加数据类型相关的容错数量。如果设置为 *N*，那么 TiDB Lightning 允许数据源中出现 *N* 个错误，而且会跳过这些错误，一旦超过这个错误数就会退出。默认值为 0，表示不允许出现错误。
+
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-error-resolution.md:21]:
```suggestion

这些错误会被记录到数据库中。在导入完成后，你可以查看数据库中的数据，手动进行处理。请参见 [错误报告](#错误报告)。

```

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 8 above)
Comment (hfxsd) [reply tidb-lightning/tidb-lightning-error-resolution.md:14]: 我补充了一些内容，请看下呢 @TomShawn

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 20 above)
Comment (TomShawn) [reply tidb-lightning/tidb-lightning-error-resolution.md:50]: LGTM

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 19 above)
Comment (TomShawn) [reply tidb-lightning/tidb-lightning-error-resolution.md:76]: LGTM

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 21 above)
Comment (TomShawn) [reply tidb-lightning/tidb-lightning-error-resolution.md:48]: LGTM

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 22 above)
Comment (gozssky) [reply tidb-lightning/tidb-lightning-error-resolution.md:21]: LGTM

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 21 above)
Comment (gozssky) [reply tidb-lightning/tidb-lightning-error-resolution.md:48]: LGTM

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 20 above)
Comment (gozssky) [reply tidb-lightning/tidb-lightning-error-resolution.md:50]: LGTM

File: tidb-lightning/tidb-lightning-error-resolution.md (hunk 19 above)
Comment (gozssky) [reply tidb-lightning/tidb-lightning-error-resolution.md:76]: LGTM

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+---
+
+# TiDB Lightning 错误处理功能
+
+从 TiDB 5.4.0 开始，你可以配置 TiDB Lightning 以跳过诸如无效类型转换、唯一键冲突等错误，让导入任务持续进行，就如同出现错误的行数据不存在一样。你可以依据生成的报告，手动修复这些错误。该功能适用于以下场景：
+
+- 要导入的数据有少许错误
+- 手动定位错误比较困难
+- 如果遇到错误就重启 TiDB Lightning，代价太大
+
```
[the remark this answers is not in the document]
Comment (gozssky) [reply tidb-lightning/tidb-lightning-error-resolution.md:13]: maybe `duplicate-resolution` -> `tikv-importer.duplicate-resolution`?

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 24]
```diff
@@ -0,0 +1,316 @@
...
+    raw_row     mediumblob NOT NULL,
+    KEY (task_id, table_name)
+);
+```
+
+**syntax_error_v1** 记录文件中的语法错误。目前尚未生效。
+
+**type_error_v1** 记录由 `max-error` 配置项管理的所有[类型错误 (Type error)](#类型错误-type-error)。每个错误一行。
+
+**conflict_error_v1** 记录所有[后端中的唯一键/主键冲突](#解决重复问题)。每对冲突有两行。
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:125]:
```suggestion
**conflict_error_v1** 记录所有[后端中的唯一键/主键冲突](#local-backend-模式下解决重复问题)。每对冲突有两行。
```

File: tidb-lightning/tidb-lightning-error-resolution.md [hunk 25]
```diff
@@ -0,0 +1,316 @@
...
+你可以使用 `lightning.task-info-schema-name` 配置更改数据库名称。
+
+{{< copyable "" >}}
+
+```toml
+[lightning]
+task-info-schema-name = 'lightning_task_info'
+```
+
+
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-error-resolution.md:84]:
```suggestion
```

**Review by TomShawn — APPROVED** [review]

File: tidb-lightning/tidb-lightning-configuration.md [hunk 26]
```diff
@@ -114,6 +114,12 @@ addr = "172.16.31.10:8287"
 # - ignore：保留已有数据，忽略新数据
 # - error：中止导入并报错
 # on-duplicate = "replace"
+# 当后端模式为 'local' 时，设置是否检测和解决重复的记录（唯一键冲突）。
+# 目前支持三种解决方法：
+#  - record: 仅将重复记录添加到目的 TiDB 中的 `lightning_task_info.conflict_error_v1` 表中。注意，该方法要求目的 TiKV 的版本为 v5.2.0 或更新版本。如果版本过低，则会启用下面的 'none' 模式。
+#  - none: 不检测重复记录。该模式是三种模式中性能最佳的，但是可能会导致目的 TiDB 中出现数据不一致的情况。
+#  - remove: 记录所有的重复记录，和 'record' 模式相似。但是会删除所有的重复记录，以确保目的 TiDB 中的数据状态保持一致。
```
Comment (TomShawn) [inline tidb-lightning/tidb-lightning-configuration.md:121]: 记录所有的重复记录又删除所有的重复记录？是不是前后矛盾呢？

Comment (hfxsd) [reply tidb-lightning/tidb-lightning-configuration.md:121]: 准确来说是 record all duplicate records to task info database 和 remove all duplicate records from the target table。

**Pull request merged**